### PR TITLE
fix(dialect-clickhouse): Support create database

### PR DESCRIPTION
**Pull request by germainlefebvre4**
### Brief summary of the change made

Features:
- Support create database
  - `Atomic`: https://clickhouse.com/docs/en/engines/database-engines/atomic
  - `Lazy`: https://clickhouse.com/docs/en/engines/database-engines/lazy
  - `Replicated`: https://clickhouse.com/docs/en/engines/database-engines/replicated
  - `PostgreSQL`: https://clickhouse.com/docs/en/engines/database-engines/postgresql
  - `MySQL`: https://clickhouse.com/docs/en/engines/database-engines/mysql
  - `SQLite`: https://clickhouse.com/docs/en/engines/database-engines/sqlite
  - `MaterializedPostgreSQL`: https://clickhouse.com/docs/en/engines/database-engines/materialized-postgresql
  - `MaterializedMySQL`: https://clickhouse.com/docs/en/engines/database-engines/materialized-mysql

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - ~`.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.~
  - ~`.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).~
  - [x] Generate tests cases fixtures in `test/fixtures/dialects/clickhouse/*.yml`.
  - [x] Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/dialects/dialect_clickhouse.py [hunk 1]
```diff
@@ -564,12 +564,42 @@ class CreateMaterializedViewStatementSegment(BaseSegment):
...
+        "DATABASE",
+        Ref("IfNotExistsGrammar", optional=True),
+        Ref("DatabaseReferenceSegment"),
+        Sequence(
+            "ON",
+            "CLUSTER",
+            Ref("ExpressionSegment"),
+            optional=True,
+        ),
+        Ref("EngineSegment"),
```
Comment (konnectr) [inline src/sqlfluff/dialects/dialect_clickhouse.py:587]: 1. I don't think you need to use an engine for tables
2. It is also optional

Comment (germainlefebvre4) [reply src/sqlfluff/dialects/dialect_clickhouse.py:587]: Good point!

PR still in WIP.

File: src/sqlfluff/dialects/dialect_clickhouse.py [hunk 2]
```diff
@@ -564,12 +564,42 @@ class CreateMaterializedViewStatementSegment(BaseSegment):
...
+    )
+
+
 class StatementSegment(ansi.StatementSegment):
     """Overriding StatementSegment to allow for additional segment parsing."""
 
     match_grammar = ansi.StatementSegment.match_grammar
     parse_grammar = ansi.StatementSegment.parse_grammar.copy(
         insert=[
+            Ref("CreateDatabaseStatementSegment"),
```
Comment (WittierDinosaur) [inline src/sqlfluff/dialects/dialect_clickhouse.py:602]: you don't need this insert as it already exists in ANSI

**Comment by konnectr** [comment]
` echo "CREATE DATABASE t1_test ENGINE = Atomic" | sqlfluff parse --dialect clickhouse -   `
```
==== parsing violations ====
L:   1 | P:  24 |  PRS | Line 1, Position 24: Found unparsable section: ' ENGINE = Atomic'
```
https://fiddle.clickhouse.com/83abaaab-9fe2-46b5-a221-ccdb35891000

**Comment by konnectr** [comment]
Clickhouse have database engine https://clickhouse.com/docs/en/engines/database-engines and deprecated engine `Ordinary`

**Comment by germainlefebvre4** [comment]
> Clickhouse have database engine https://clickhouse.com/docs/en/engines/database-engines and deprecated engine `Ordinary`

I haven't found the `Ordinary` database engine you're talking about. Nevertheless I tested resiliency of all support (and experimental) Database Engine as describe in the PR description.

**Comment by konnectr** [comment]
I don't understand what we are trying to achieve? Now a random word after ENGINE is accepted and there is no error. How should this be regulated?

`echo "CREATE DATABASE t1_test ENGINE = asdasdad" | sqlfluff parse --dialect clickhouse - `

File: src/sqlfluff/dialects/dialect_clickhouse.py [hunk 3]
```diff
@@ -303,7 +303,107 @@ class EngineSegment(BaseSegment):
...
+                Sequence(
+                    "SAMPLE",
+                    "BY",
+                    Ref("ExpressionSegment"),
+                    optional=True,
+                ),
+                Sequence(
+                    "SETTINGS",
+                    Delimited(
+                        AnyNumberOf(
```
Comment (WittierDinosaur) [inline src/sqlfluff/dialects/dialect_clickhouse.py:338]: This AnyNumberOf isn't needed, as Delimited covers that behaviour

File: src/sqlfluff/dialects/dialect_clickhouse.py [hunk 4]
```diff
@@ -303,7 +303,107 @@ class EngineSegment(BaseSegment):
...
+class DatabaseEngineFunctionSegment(BaseSegment):
+    """A ClickHouse `ENGINE` clause function.
+
+    With this segment we attempt to match all possible engines.
+    """
+
+    type = "engine_function"
+    match_grammar: Matchable = Sequence(
+        Sequence(
+            # Ref(
```
Comment (WittierDinosaur) [inline src/sqlfluff/dialects/dialect_clickhouse.py:366]: what's with the commented out code?

File: src/sqlfluff/dialects/dialect_clickhouse_keywords.py [hunk 5]
```diff
@@ -171,4 +172,13 @@
     "WHERE",
     "WITH",
     "YEAR",
+    # Database Engine
+    "Atomic",
```
Comment (WittierDinosaur) [inline src/sqlfluff/dialects/dialect_clickhouse_keywords.py:176]: Can we please follow the convention of keywords being uppercase?

File: src/sqlfluff/dialects/dialect_clickhouse.py [hunk 6]
```diff
@@ -303,7 +303,107 @@ class EngineSegment(BaseSegment):
...
+            #     "FunctionNameSegment",
+            #     exclude=OneOf(
+            #         Ref("DatePartFunctionNameSegment"),
+            #         Ref("ValuesClauseSegment"),
+            #     ),
+            # ),
+            OneOf(
+                "Atomic",
```
Comment (WittierDinosaur) [inline src/sqlfluff/dialects/dialect_clickhouse.py:374]: Can we please follow the convention of uppercase keywords here too?

File: src/sqlfluff/dialects/dialect_clickhouse.py [hunk 7]
```diff
@@ -564,13 +664,78 @@ class CreateMaterializedViewStatementSegment(BaseSegment):
...
+                optional=True,
+            ),
+        ),
+        AnyNumberOf(
+            "TABLE",
+            "OVERRIDE",
+            Ref("TableReferenceSegment"),
+            Bracketed(
+                Delimited(
+                    OneOf(
```
Comment (WittierDinosaur) [inline src/sqlfluff/dialects/dialect_clickhouse.py:720]: Delimited has this OneOf functionality, so here it isn't needed

File: src/sqlfluff/dialects/dialect_clickhouse.py [hunk 8]
```diff
@@ -303,7 +303,107 @@ class EngineSegment(BaseSegment):
...
+            Ref("TableEngineFunctionSegment"),
+            AnySetOf(
+                Sequence(
+                    "ORDER",
+                    "BY",
+                    OneOf(
+                        Ref("BracketedColumnReferenceListGrammar"),
+                        Ref("ColumnReferenceSegment"),
+                    ),
+                    optional=True,
```
Comment (WittierDinosaur) [inline src/sqlfluff/dialects/dialect_clickhouse.py:315]: The optionals in these sequences aren't needed, as it's covered in AnySetOf

File: src/sqlfluff/dialects/dialect_clickhouse.py [hunk 9]
```diff
@@ -564,13 +664,78 @@ class CreateMaterializedViewStatementSegment(BaseSegment):
...
+            Sequence(
+                "ON",
+                "CLUSTER",
+                Ref("ExpressionSegment"),
+                optional=True,
+            ),
+            Ref("DatabaseEngineSegment", optional=True),
+            Sequence(
+                "COMMENT",
+                Ref("ExpressionSegment"),
```
Comment (WittierDinosaur) [inline src/sqlfluff/dialects/dialect_clickhouse.py:691]: This doesn't need to be an expression

File: src/sqlfluff/dialects/dialect_clickhouse.py [hunk 10]
```diff
@@ -564,13 +664,78 @@ class CreateMaterializedViewStatementSegment(BaseSegment):
...
+                optional=True,
+            ),
+            Sequence(
+                "SETTINGS",
+                Delimited(
+                    AnyNumberOf(
```
Comment (WittierDinosaur) [inline src/sqlfluff/dialects/dialect_clickhouse.py:697]: This AnyNumberOf isn't needed

File: src/sqlfluff/dialects/dialect_clickhouse.py [hunk 11]
```diff
@@ -564,13 +664,78 @@ class CreateMaterializedViewStatementSegment(BaseSegment):
...
+    match_grammar = Sequence(
+        "CREATE",
+        "DATABASE",
+        Ref("IfNotExistsGrammar", optional=True),
+        Ref("DatabaseReferenceSegment"),
+        AnySetOf(
+            Sequence(
+                "ON",
+                "CLUSTER",
+                Ref("ExpressionSegment"),
```
Comment (WittierDinosaur) [inline src/sqlfluff/dialects/dialect_clickhouse.py:685]: This doesn't need to be an Expression

File: src/sqlfluff/dialects/dialect_clickhouse.py (hunk 3 above)
Comment (germainlefebvre4) [reply src/sqlfluff/dialects/dialect_clickhouse.py]: Fixed in commit `f56765e`.

File: src/sqlfluff/dialects/dialect_clickhouse.py (hunk 4 above)
Comment (germainlefebvre4) [reply src/sqlfluff/dialects/dialect_clickhouse.py]: Fixed in commit `f56765e`.

File: src/sqlfluff/dialects/dialect_clickhouse_keywords.py (hunk 5 above)
Comment (germainlefebvre4) [reply src/sqlfluff/dialects/dialect_clickhouse_keywords.py]: Fixed in commit `f56765e`.

File: src/sqlfluff/dialects/dialect_clickhouse.py (hunk 6 above)
Comment (germainlefebvre4) [reply src/sqlfluff/dialects/dialect_clickhouse.py]: Fixed in commit `f56765e`.

File: src/sqlfluff/dialects/dialect_clickhouse.py (hunk 7 above)
Comment (germainlefebvre4) [reply src/sqlfluff/dialects/dialect_clickhouse.py]: Fixed in commit `f56765e`.

File: src/sqlfluff/dialects/dialect_clickhouse.py (hunk 8 above)
Comment (germainlefebvre4) [reply src/sqlfluff/dialects/dialect_clickhouse.py]: Fixed in commit `f56765e`.

File: src/sqlfluff/dialects/dialect_clickhouse.py (hunk 10 above)
Comment (germainlefebvre4) [reply src/sqlfluff/dialects/dialect_clickhouse.py:683]: Fixed in commit `f56765e`.

File: src/sqlfluff/dialects/dialect_clickhouse.py (hunk 11 above)
Comment (germainlefebvre4) [reply src/sqlfluff/dialects/dialect_clickhouse.py:671]: How may I detect a `CLUSTER` pattern? (also for `ON COMMENT`)

Comment might be string and unquoted single word .

File: src/sqlfluff/dialects/dialect_clickhouse.py (hunk 9 above)
Comment (germainlefebvre4) [reply src/sqlfluff/dialects/dialect_clickhouse.py:677]: How may I detect a `COMMENT` pattern? (also for `ON CLUSTER`)

Comment might be commented string and single unquoted word.

File: src/sqlfluff/dialects/dialect_clickhouse.py (hunk 11 above)
Comment (WittierDinosaur) [reply src/sqlfluff/dialects/dialect_clickhouse.py:671]: I think what you're after is `SingleIdentifierGrammar`

**Comment by germainlefebvre4** [comment]
I updated the ANSI dialect and included "Single Quoted" to targeted Identifiers.

**Review by WittierDinosaur — APPROVED** [review]
LGTM

**Review by konnectr — APPROVED** [review]

**Comment by WittierDinosaur** [comment]
Merge conflicts

**Comment by germainlefebvre4** [comment]
Working on it

**Pull request merged**